### ISSUE TITLE
Cannot read uploaded file

### ISSUE THREAD

**Issue by leledumbo** [fog/fog-softlayer#47]
We're using the following method to connect to softlayer object storage service:

```
def method(a,b)
  sl = Fog::Storage.new(
    provider: 'softlayer',
    softlayer_storage_account: '<our account>',
    softlayer_username: '<our username>',
    softlayer_api_key: '<our api key>',
    softlayer_cluster: '<our cluster>',
  )

  dirkey = "<key>"

  dir = sl.directories.get(dirkey)
  if dir.nil?
    dir = sl.directories.create(:key => dirkey) 
    dir.save
  end

  begin
    dirname = File.basename(File.dirname(b))
    filename = File.basename(filepath)
    content  = File.open(filepath).read

    filekey = "#{a}/#{dirname}/#{b}"

    file = dir.files.create(:key => filekey, :body => content)
    file.save
  end
end
```

I can confirm that the file is successfully uploaded because I can see it in the storage. However, when I want to access it from URL I get:

> Unauthorized
> This server could not verify that you are authorized to access the document you requested.

This doesn't happen when we use the PHP bindings. Debugging both libs, it seems that I need to add

```ruby
X-CONTAINER-READ: .r:*
```

to the headers. I tried to put it in:

```ruby
connection_options: { 'X-CONTAINER-READ' => '.r:*' }
```
when calling `Fog::Storage.new` above, but the result is the same. What am I supposed to do?

**Comment by cphrmky** [comment]
@leledumbo ,

Are you using a signed URL when you try to access the file/object?  Any directories created by the code you've posted above are going to have the default access settings, which are not entirely public.

Here's any example of how to use the method in `Fog::Storage` to generate a signed URL for accessing the object.

```ruby
file = @sl.directories.get('a-container').files.get('data.txt')
file.url(Time.now + 300) # url expires in 5 minutes
# => "https://dal05.objectstorage.softlayer.net:443/v1/AUTH_1a1a1a1a-1a1a-1a1a-1a1a-1a1a1a1a1a1a/a-container/data.txt?temp_url_sig=1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a&temp_url_expires=1401901023"
```

If you wanted to create a "forever" URL for an object, you can create a URL that expires in 100 years or something like so:  

```ruby
file.url(Time.now + 3153600000) # url expires in 100 years
```

Please give this a shot and let me know if you're still stuck, thanks!

**Comment by leledumbo** [comment]
> If you wanted to create a "forever" URL for an object, you can create a URL that expires in 100 years or something like so:

That was what I had in mind but my leader refuses the idea since it's actually possible to enable public cdn access, which can be done by sending `X-CONTAINER-READ: .r:*` header when creating the container (I think that's what `public_url` method should do). We had a PHP code that does this and it works as expected, but our background worker is written in Ruby and we don't want to change the language because it contains 4 other workers already.

**Comment by cphrmky** [comment]
@leledumbo ,

Gotcha, there are valid reasons for not wanting to use a "forever" signed URL.  I'm labeling this as an enhancement request.  What I'll end up doing is adding a flag so that you can do this:

```ruby
dir = sl.directories.create(:key => dirkey, :public => true) 
dir.save
```

It will default to `false` so as to maintain backwards compatibility.  

Not sure when I'll circle around to getting this implemented at this point, do you need this urgently?  Can you manually set your created directory/directories to public in the interim?

Let me know if/how urgent it is and I can prioritize it a little better.

**Comment by leledumbo** [comment]
> do you need this urgently? Can you manually set your created directory/directories to public in the interim?

Not anymore, we fallback to Ruby calling PHP for now. We can't do that manually as this code will be called automatically by another code in our framework, and it will happen through user action, which we can't control.

A pure Ruby solution is better in the future so that less cross language maintenance required, though. So we're still expecting the enhancement.

**Comment by cphrmky** [comment]
@leledumbo thanks for the update.  I'll be sure to get this prioritized and done hopefully in the next couple weeks.

**Comment by cphrmky** [comment]
@leledumbo this has been fixed in [v0.4.5](https://rubygems.org/gems/fog-softlayer/versions/0.4.5).  You'll need to modify the code you're running a bit to use `.new` rather than `.create`, and to set the `public` attribute like so:

```ruby
dir = sl.directories.get(dirkey)
if dir.nil?
  dir = sl.directories.new(:key => dirkey)
  dir.public = true
  dir.save # => true

  dir.public? # => true
end
```

**Comment by leledumbo** [comment]
Though I wonder why you need to invent a new method instead of just providing an attribute, it's fine for me. Thanks for the implementation
